essential-scala/structural-recursion.md

```diff
@@ -1,6 +1,198 @@
+
+{:teacher:} Point out how `null` or `exceptions` break this rule of *only one implementation based on type definition* and how this shows that these kind of side effects/unexpected behaviour is bad. We don't like suprises.
+
+## Pattern matching
+
+Pattern matching is a way to work with ADT, making decisions based on the shape of the data. This is different than calling methods or accessing fields on a type, and more powerful at the same time.
+
+The best way to see the difference is to start with an example. We will use this ADT we built in the previous lesson:
+
+```scala
+sealed trait Attendee
+final case class Student(name: String, goal: String) extends Attendee
+final case class Mentor(name: String, yearsOfExperience: Int) extends Attendee
+```
+
+Let's imagine that we want to know at which time an `Attendee` to Scala Bridge must be at the location the event takes place. This time will be different for `Student` and `Mentor`. 
+
+If you are used to OO, you could imagine an implementation to solve this problem by using a field in `Attendee`:
+
+```scala
+sealed trait Attendee {
+   val timeAtLocation: LocalDateTime
+}
+final case class Student(name: String, goal: String, timeAtLocation: LocalDateTime) extends Attendee
+final case class Mentor(name: String, yearsOfExperience: Int, timeAtLocation: LocalDateTime) extends Attendee
+```
+
+but this has some unwanted consequences. We must provide the time when we create an instance of `Attendee` and we will have to modify existing code currently using these `case classes`. And what about time for multiple locations? 
+
+A more FP solution would be to implement the following function:
+
+```scala
+def timeAtLocation(attendee: Attendee): LocalDateTime = ???
+```
+
+which we can call for each `Attendee` for whom we want to check the `timeAtLocation`. But, how do we know if an `Attendee` is an `Student` or a `Mentor`? If you are used to OO you may be thinking on `casting` the type (if you don't know what this means, that's good as it is not a good practice in Scala!). In Scala the solution is `Pattern Matching`:
+
+```scala 
+def timeAtLocation(attendee: Attendee): LocalDateTime = attendee match {
+   case Student(name, goal) =>  LocalDateTime.of(2019, Month.APRIL, 4, 18, 00)
+   case Mentor(name, yearsOfExperience) => LocalDateTime.of(2019, Month.APRIL, 4, 17, 30)
+}
+```
+
+By using the `match` and `case` keyword we create a piece of code that changes behaviour based on the `shape` of the ADT we provide to the function. Note that we have access to the fields in the ADT, so we can use them in our logic. For example, we could print to console details about the `Attendee` as we make the decision: 
+
+```scala 
+def timeAtLocation(attendee: Attendee): LocalDateTime = attendee match {
```

Comment:
It's a picky thing but I think that timeAtLocation is actually the arrivalTime and therefore should be renamed as such. For me if it is TimeAtLocation I am expecting a duration rather than a time.

Reply:
ah, that's a very good point, will fix